### PR TITLE
Implement lattice transform to upper-triangular matrix for ASE Nose-Hoover thermostat

### PR DESCRIPTION
**Pull request by tsihyoung**
0db053c Implement transformation of lattice to upper-triangular matrix for ASE Nose-Hoover thermostat
dfd8bc4 Add verbose flag and explanation to upper_triangular_cell
1beaf1b fix ASE NPT does not accept externalstress=None
f22adc1 Add tests for Nose-Hoover NVT and NPT MD
a6480fd Change MD test numerical tolerance
a976874 use numpy.testing.assert_allclose over np.isclose().all() for better error messages on failing tests
cfba1c9 tweak upper_triangular_cell() doc str

### CODE REVIEW

File: chgnet/model/dynamics.py [hunk 1]
```diff
@@ -651,6 +653,29 @@ def set_atoms(self, atoms: Atoms):
...
+        if not NPT._isuppertriangular(self.atoms.get_cell()):
+            a, b, c, alpha, beta, gamma = self.atoms.cell.cellpar()
+            ang = np.radians((alpha, beta, gamma))
+            sina, sinb, sing = np.sin(ang)
+            cosa, cosb, cosg = np.cos(ang)
+            cosp = (cosg - cosa * cosb) / (sina * sinb)
+            if cosp > 1.0:
+                cosp = 1.0
+            elif cosp < -1.0:
+                cosp = -1.0
```
Comment (janosh) [inline chgnet/model/dynamics.py:667]: How about `np.clip(cosp, 1, -1)` here?

File: chgnet/model/dynamics.py [hunk 2]
```diff
@@ -651,6 +653,29 @@ def set_atoms(self, atoms: Atoms):
         self.dyn.atoms = atoms
         self.dyn.atoms.calc = calculator
 
+    def upper_triangular_cell(self):
+        """Transform to upper triangular cell."""
```
Comment (janosh) [inline chgnet/model/dynamics.py:657]: For posterity's sake, can you add a comment to this doc string explaining what this is useful for.

File: chgnet/model/dynamics.py [hunk 3]
```diff
@@ -651,6 +653,29 @@ def set_atoms(self, atoms: Atoms):
...
+            sinp = np.sqrt(1.0 - cosp**2.0)
+
+            new_basis = [
+                (a * sinb * sinp, a * sinb * cosp, a * cosb),
+                (0.0, b * sina, b * cosa),
+                (0.0, 0.0, c),
+            ]
+
+            self.atoms.set_cell(new_basis, scale_atoms=True)
+            print("Transformed to upper triangular unit cell.", flush=True)
```
Comment (janosh) [inline chgnet/model/dynamics.py:677]: This should not always print. Let's add a `verbose` flag.

**Review by janosh — CHANGES_REQUESTED** [review]
Thanks @tsihyoung! I left some comments and we need a test for this.

**Comment by tsihyoung** [comment]
> 

Thanks @janosh for reviewing the codes.

Do we need a test for NPT dynamics or for the transformation itself? If the latter is the case, the test might be not relevant enough to be included in this repository.

**Comment by janosh** [comment]
Could be I'm missing something. My understanding is you're addressing https://github.com/CederGroupHub/chgnet/pull/67#issuecomment-1709170273 from @BowenD-UCB. In which case this enables MD for systems that weren't supported before? If so, testing such a system would be good.

**Comment by BowenD-UCB** [comment]
> > Could be I'm missing something. My understanding is you're addressing [#67 (comment)](https://github.com/CederGroupHub/chgnet/pull/67#issuecomment-1709170273) from @BowenD-UCB. In which case this enables MD for systems that weren't supported before? If so, testing such a system would be good.
> 
> Yes, that is what I meant. I will add testings for Nose-Hoover NVT and NPT dynamics.

Thanks for helping addressing this issue!
Yes, the test should include: 
1.showing a non-upper triangular structure being transformed into upper-triangular structure
2. verify they're consistent by using pymatgen structure matcher
3. 3 and then running noose-hoover (both nvt and npt) on it

**Comment by tsihyoung** [comment]
> > > Could be I'm missing something. My understanding is you're addressing [#67 (comment)](https://github.com/CederGroupHub/chgnet/pull/67#issuecomment-1709170273) from @BowenD-UCB. In which case this enables MD for systems that weren't supported before? If so, testing such a system would be good.
> > 
> > 
> > Yes, that is what I meant. I will add testings for Nose-Hoover NVT and NPT dynamics.
> 
> Thanks for helping addressing this issue! Yes, the test should include:
> 
> 1. showing a non-upper triangular structure being transformed into upper-triangular structure
> 2. verify they're consistent by using pymatgen structure matcher
> 3. and then running noose-hoover (both nvt and npt) on it

I have added the required tests. By the way, although the default value of `externalstress` in ASE NPT class is `None`, it will cause `ValueError` as `set_stress` has not considered the case where `externalstress is None`. This has also been fixed.

**Review by BowenD-UCB — APPROVED** [review]
This looks good so far

File: tests/test_md.py [hunk 4]
```diff
@@ -24,6 +27,7 @@
 
 relaxer = StructOptimizer()
 structure = Structure.from_file(f"{ROOT}/examples/mp-18767-LiMnO2.cif")
+structure_2 = Structure.from_file(f"{ROOT}/examples/mp-1175469-Li9Co7O16.cif")
```
Comment (BowenD-UCB) [inline tests/test_md.py:30]: Why is there a need to include LCO for Nose-Hoover test?

Comment (tsihyoung) [reply tests/test_md.py:30]: > Why is there a need to include LCO for Nose-Hoover test?

Because the cell matrix of LiMnO2 is diagonal.

Comment (BowenD-UCB) [reply tests/test_md.py:30]: This is makes sense. Another way is to use rotation and non-diagonal supercell expansion to create a non-diagonal LiMnO2 cell

**Review by janosh — APPROVED** [review]
Looks great, thanks @tsihyoung! 👍

**Pull request merged**